### ISSUE TITLE
How are optimisations handled?

### ISSUE THREAD

**Issue by zenna** [Keno/Cxx.jl#116]
Apologies in advance if this is not the appropriate place for this kind of query.

I am very naive of how Cxx actually works, but it seems that in contrast to the normal C++ compilation process, the optimisation opportunities are much fewer and more limited  So I'm interested to understand how optimisations are handled in general, but to be concrete I have a set of questions:

- Does Cxx currently do any optimisation?
- Is this something that could be a (runtime or compile time) option (in the same way I choose optimisations for g++ or clang)?
- Will Cxx be always inherently slower than precompiled C++ code?
- Will a compiled C++ shared library be more performant than a header library?
- If I care about performance should I write a C wrapper for my C++ library and  use ccall as is conventionally done (and avoid Cxx all together)?

**Comment by Keno** [comment]
> Does Cxx currently do any optimisation?

It runs the same optimizations as julia, because the LLVM IR gets inlined into the julia function. This does mean we lose a bit of LLVM level inlining, because julia modules are smaller, and we don't actually bother running the inlining pass.

> Is this something that could be a (runtime or compile time) option (in the same way I choose optimisations for g++ or clang)?

It'll be affected by the julia optimization flags.

> Will Cxx be always inherently slower than precompiled C++ code?

Yes, because of inlining, same with regular julia code though.

> Will a compiled C++ shared library be more performant than a header library?

They are usually used for different purposes, so it depends what you mean here.

> If I care about performance should I write a C wrapper for my C++ library and use ccall as is conventionally done (and avoid Cxx all together)?

No, for all cases that such overhead matters, you get the biggest benefit by inlining the code into the calling function. Cxx does that for you, a ccall does not.

**Comment by Mike43110** [comment]
I was wondering - would it be possible at a later stage to change the Julia optimisation flags on a per-program basis?

Or, more ambitiously on say a more fine grained level such as at the module level to gain the LLVM inlining pass? If more fine grained optimisations could be done would it then be possible to compile different functions with different levels?

This may belong on the main Julia page but why I ask this is that it could be a -nice to haves- with pre-compilation once completed. Albeit I think that backtraces may be affected. That I am not too sure on.

Just wondering if it would even be possible at any level. Per-module would be quite nice imo for packages.

**Comment by Keno** [comment]
Yes, this would be theoretically possible, though there'd be quite a bit of design work involved to pass this through from syntax to codegen.
